core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+ * }</pre>
+ * And then a successful transfer may output something similar to:
+ * <pre>
+ * Request initiated...
+ * |                    | 0.0%
+ * |==                  | 12.5%
+ * |=====               | 25.0%
+ * |=======             | 37.5%
+ * |==========          | 50.0%
+ * |============        | 62.5%
+ * |===============     | 75.0%
+ * |=================   | 87.5%
+ * |====================| 100.0%
+ * Request execution successful!
+ * </pre>
+ */
+@SdkPublicApi
+public interface ProgressListener {
+
+    /**
+     * This method is called right after a request object is marshalled and ready to be sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestPrepared#request()}</li>
+     *     <li>{@link Context.RequestPrepared#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestPrepared(Context.RequestPrepared context) {
+    }
+
+    /**
+     * This method is called after the request transaction is initiated, i.e. request header is sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestHeaderSent#request()}</li>
+     *     <li>{@link Context.RequestHeaderSent#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestHeaderSent(Context.RequestHeaderSent context) {
+    }
+
+    /**
+     * This method is called with any additional payload bytes sent; it may be called many times per request, depending on the
+     * payload size.
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestBytesSent#request()}</li>
+     *     <li>{@link Context.RequestBytesSent#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestBytesSent(Context.RequestBytesSent context) {
+    }
+
+    /**
+     * The service returns the response headers
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ResponseHeaderReceived#request()}</li>
+     *     <li>{@link Context.ResponseHeaderReceived#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void responseHeaderReceived(Context.ResponseHeaderReceived context) {
+    }
+
+    /**
+     * Additional bytes received
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ResponseBytesReceived#request()}</li>
+     *     <li>{@link Context.ResponseBytesReceived#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void responseBytesReceived(Context.ResponseBytesReceived context) {
+    }
+
+    /**
+    * Additional bytes received
+    * <p>
+    * Available context attributes:
+    * <ol>
+    *     <li>{@link Context.ResponseBytesReceived#request()}</li>
+    *     <li>{@link Context.ResponseBytesReceived#progressSnapshot()}</li>
+    * </ol>
+    */
+    default void attemptFailureResponseBytesReceived(Context.AttemptFailureResponseBytesReceived context) {
+    }
```

Comment:
Can we describe the differences between responseBytesReceived and attemptFailureResponseBytesReceived here?